server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java

```diff
@@ -119,35 +119,28 @@ public Mappings getMappings(String user, Domain domain, int mappingLimit) throws
                 MappingsImpl.Builder mappings = MappingsImpl.builder();
 
                 for (String target : targetMappings.asStrings()) {
-                    if (Mapping.detectType(target).equals(Type.Regex)) {
-                        try {
-                            target = RecipientRewriteTableUtil.regexMap(new MailAddress(user, domain.asString()), target);
-                        } catch (PatternSyntaxException | ParseException e) {
-                            LOGGER.error("Exception during regexMap processing: ", e);
-                        }
-                    } else if (Mapping.detectType(target).equals(Type.Domain)) {
-                        target = user + "@" + Type.Domain.withoutPrefix(target);
-                    }
+                    Type type = Mapping.detectType(target);
+                    Optional<String> maybeAddressWithMappingApplied = applyMapping(user, domain, target, type);
 
-                    if (target == null) {
+                    if (!maybeAddressWithMappingApplied.isPresent()) {
                         continue;
                     }
-
-                    String buf = "Valid virtual user mapping " + user + "@" + domain.name() + " to " + target;
+                    String addressWithMappingApplied = maybeAddressWithMappingApplied.get();
+                    String buf = "Valid virtual user mapping " + user + "@" + domain.name() + " to " + maybeAddressWithMappingApplied;
```

Comment:
concat of an Optional with implicit toString ?

Reply:
maybeAddressWithMappingApplied is a String....